lib/vector-core/src/event/metric/mod.rs

```diff
@@ -0,0 +1,1008 @@
+        &mut self.metadata
+    }
+
+    /// Gets a reference to the name of this metric.
+    ///
+    /// The name of the metric does not include the namespace or tags.
+    #[inline]
+    pub fn name(&self) -> &str {
+        &self.series.name.name
+    }
+
+    /// Gets a reference to the namespace of this metric, if it exists.
+    #[inline]
+    pub fn namespace(&self) -> Option<&str> {
+        self.series.name.namespace.as_deref()
+    }
+
+    /// Takes the namespace out of this metric, if it exists, leaving it empty.
+    #[inline]
+    pub fn take_namespace(&mut self) -> Option<String> {
+        self.series.name.namespace.take()
+    }
+
+    /// Gets a reference to the tags of this metric, if they exist.
+    #[inline]
+    pub fn tags(&self) -> Option<&MetricTags> {
+        self.series.tags.as_ref()
+    }
+
+    /// Gets a reference to the timestamp of this metric, if it exists.
+    #[inline]
+    pub fn timestamp(&self) -> Option<DateTime<Utc>> {
+        self.data.timestamp
+    }
+
+    /// Gets a reference to the value of this metric.
+    #[inline]
+    pub fn value(&self) -> &MetricValue {
+        &self.data.value
+    }
+
+    /// Gets the kind of this metric.
+    #[inline]
+    pub fn kind(&self) -> MetricKind {
+        self.data.kind
+    }
+
+    /// Decomposes a `Metric` into its individual parts.
+    #[inline]
+    pub fn into_parts(self) -> (MetricSeries, MetricData, EventMetadata) {
+        (self.series, self.data, self.metadata)
+    }
+
+    /// Creates a `Metric` directly from the raw components of another metric.
+    #[inline]
+    pub fn from_parts(series: MetricSeries, data: MetricData, metadata: EventMetadata) -> Self {
+        Self {
+            series,
+            data,
+            metadata,
+        }
+    }
+
+    /// Consumes this metric, returning it as an absolute metric.
+    ///
+    /// If the metric was already absolute, nothing is changed.
+    #[must_use]
+    pub fn into_absolute(self) -> Self {
+        Self {
+            series: self.series,
+            data: self.data.into_absolute(),
+            metadata: self.metadata,
+        }
+    }
+
+    /// Consumes this metric, returning it as an incremental metric.
+    ///
+    /// If the metric was already incremental, nothing is changed.
+    #[must_use]
+    pub fn into_incremental(self) -> Self {
+        Self {
+            series: self.series,
+            data: self.data.into_incremental(),
+            metadata: self.metadata,
+        }
+    }
+
+    /// Creates a new metric from components specific to a metric emitted by `metrics`.
+    #[allow(clippy::cast_precision_loss)]
+    pub fn from_metric_kv(key: &metrics::Key, handle: &Handle) -> Self {
+        let value = match handle {
+            Handle::Counter(counter) => MetricValue::Counter {
+                // NOTE this will truncate if `counter.count()` is a value
+                // greater than 2**52.
+                value: counter.count() as f64,
```

Comment:
I don't love silent truncation like this, even if I admit that in practice it will probably be rare. I also wouldn't want to crash vector either. 

I'm actually not sure that we really document what the internal supported sizes for things are in vector, so I suppose in that context this behavior is as good as any. It's easy to find on account of the comment, that said.

Reply:
So, "truncate" is a little too out on its own here.  It's really a loss of precision i.e. the inability to distinguish between 2^53+1 and 2^53+2 because they both can't be represented in a lossless way.

So we're truncating some of the lower bits, even though technically `f64` can reach greater absolute values than `u64` can.

I do agree, on the whole though, that this is not great from a "we are as correct as we can possibly be" standpoint, although somehow things like Prometheus get away with this and they haven't been ridden out of town yet, so... 🤷🏻

Reply:
Haha, fair point.